File: app/src/main/java/com/example/budgetbuddy/entities/Transaction.java
package com.example.budgetbuddy.entities;

import java.util.Calendar;
import java.util.Locale;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import com.example.budgetbuddy.enums.RepeatDuration;
import com.example.budgetbuddy.enums.TransactionType;

/**
 * Represents a transaction in the database with an ID, amount, type, date and time, category and repeat duration.
 * <code>nextID</code> is a static long variable used to generate unique IDs. It is shared by all <code>Transaction</code> objects
 * and is incremented each time a new object is created.
 */
@Entity(tableName = "transaction")
public final class Transaction
{
    private static long nextID;                     // Unique ID generator

    @PrimaryKey
    private long id;                                // Primary key

    @ColumnInfo(name = "amount")
    private final double amount;                    // Amount field

    @ColumnInfo(name = "type")
    private final TransactionType type;             // Type field

    @ColumnInfo(name = "datetime")
    private final Calendar dateTime;                // DateTime field

    @ColumnInfo(name = "category")
    private final long categoryID;                  // CategoryID foreign key

    @ColumnInfo(name = "repeat")
    private final RepeatDuration repeatDuration;    // RepeatDuration field

    /**
     * Constructs a new Transaction object with the given parameters. Sets <code>id</code> to <code>nextID</code> and
     * increments it
     *
     * @param amount         a <code>double</code> value for the amount
     * @param type           a <code>TransactionType</code> value for the type
     * @param dateTime       a <code>Calendar</code> object for the date and time
     * @param categoryID     a <code>long</code> value for the categoryID
     * @param repeatDuration a <code>RepeatDuration</code> value for the repeat duration
     */
    public Transaction(double amount, TransactionType type, Calendar dateTime, long categoryID, RepeatDuration repeatDuration)
    {
        this.id = nextID++; // Increment the ID counter by 1
        this.amount = amount;
        this.type = type;
        this.dateTime = dateTime;
        this.categoryID = categoryID;
        this.repeatDuration = repeatDuration;
    }

    /**
     * Gets the transaction ID
     *
     * @return the transaction ID
     */
    public long getId()
    {
        return id;
    }

    /**
     * Sets the transaction ID
     *
     * @param id the transaction ID
     */
    public void setId(long id)
    {
        this.id = id;
    }

    /**
     * Gets the transaction amount
     *
     * @return the transaction amount
     */
    public double getAmount()
    {
        return amount;
    }

    /**
     * Gets the transaction type
     *
     * @return the transaction type
     */
    public TransactionType getType()
    {
        return type;
    }

    /**
     * Gets the date and time of the transaction
     *
     * @return a <code>Calendar</code> object containing the date and time
     */
    public Calendar getDateTime()
    {
        return dateTime;
    }

    /**
     * Gets the category ID of the transaction
     *
     * @return the category ID
     */
    public long getCategoryID()
    {
        return categoryID;
    }

    /**
     * Gets the repeat duration of the transaction
     *
     * @return the repeat duration
     */
    public RepeatDuration getRepeatDuration()
    {
        return repeatDuration;
    }

    /**
     * Returns a string representation of the transaction
     *
     * @return a <code>String</code> object combining all transaction data
     */
    @NonNull
    public String toString()
    {
        return String.format(Locale.getDefault(), "ID: %s, Amount: %.2f, Type: %s, Date: %s", id, amount, type, dateTime.toString());
    }

    /**
     * Compares two transactions by their <code>dateTime</code> in milliseconds
     *
     * @param o the reference object with which to compare.
     * @return true if the objects are equal, false otherwise
     */
    @Override
    public boolean equals(Object o)
    {
        if (this == o) return true; // Check if the two objects are exactly the same in memory
        if (o == null || getClass() != o.getClass())
            return false; // Check if the other object is not a Transaction or is null

        Transaction x = (Transaction) o; // Cast o to transaction- only executed if the last check succeeded

        // Compare the date and time in milliseconds
        return Double.compare(x.dateTime.getTimeInMillis(), dateTime.getTimeInMillis()) == 0;
    }

    /**
     * Generates a unique hashcode for this <code>Transaction</code> using all of its fields
     *
     * @return a <code>int</code> value representing the hashcode
     */
    @Override
    public int hashCode()
    {
        return Objects.hash(id, amount, type, dateTime, categoryID, repeatDuration);
    }

    /**
     * Gets the Transaction date and time as a String in the format hh:mm dd/MM/yyyy
     *
     * @return a <code>String</code> object combining the date and time
     */
    @NonNull
    public String getDateTimeString()
    {
        return String.format(Locale.getDefault(), "%02d:%02d %02d/%02d/%d", dateTime.get(Calendar.HOUR_OF_DAY), dateTime.get(Calendar.MINUTE), dateTime.get(Calendar.DAY_OF_MONTH), dateTime.get(Calendar.MONTH) + 1, dateTime.get(Calendar.YEAR));
    }


}

File: app/src/main/java/com/example/budgetbuddy/domain/entities/Transaction.kt
package com.example.budgetbuddy.domain.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.example.budgetbuddy.enums.RepeatDuration
import com.example.budgetbuddy.enums.TransactionType
import com.example.budgetbuddy.domain.Result
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

/**
 * Data class for Transaction objects acting as an Entity in the room database.
 * Auto-generates primary keys.
 */
@Entity(tableName = "transaction")
data class Transaction(
    @PrimaryKey(autoGenerate = true)    // Let Room handle ID generation
    val id: Long = 0,                   // kotlin does not have primitive long

    @ColumnInfo(name = "amount")
    val amount: Double,                 // nor double

    @ColumnInfo(name = "type")
    val type: TransactionType,

    @ColumnInfo(name = "datetime")
    val dateTime: Calendar,

    @ColumnInfo(name = "categoryID")
    val categoryID: Long,               // Foreign key to Category object

    @ColumnInfo(name = "repeat")
    val repeatDuration: RepeatDuration
)
{
    /**
     * Overrides toString method from base data class. Outputs transaction data
     * for logging
     */
    override fun toString(): String
    {
        return String.format(
            Locale.getDefault(),
            "ID: %s, Amount: %.2f, Type: %s, Date: %s",
            id, amount, type, dateTimeString
        )
    }

    // Define the formatting for the dateTime variable
    val dateTimeString: String
        get() = String.format(      // "custom" get setter
            Locale.getDefault(),
            "%02d:%02d %02d/%02d/%d",
            dateTime.get(Calendar.HOUR_OF_DAY),
            dateTime.get(Calendar.MINUTE),
            dateTime.get(Calendar.DAY_OF_MONTH),
            dateTime.get(Calendar.MONTH) + 1,
            dateTime.get(Calendar.YEAR)
        )
}

sealed interface TransactionError
{
    data object Empty : TransactionError
    data object BadFormat : TransactionError
    data object InvalidAmount : TransactionError
    data object NoCategory : TransactionError
    data object InvalidDateTime : TransactionError
}

enum class TransactionType
{
    INCOMING,
    OUTGOING
}

enum class RepeatDuration
{
    NEVER,
    DAILY,
    WEEKLY,
    MONTHLY,
    YEARLY
}


/**
 * Transaction amounts are taken in as strings and parsed to doubles
 * in this value class. Returns TransactionError if parsing fails
 *
 *
 *               Regex:
 *               [0-9]+               Matches if the string contains one-or-more numbers
 *               [.]                  Matches if the string contains a decimal point
 *               [0-9]{1, 2}          Matches if numeric and at most 2 d.p.
 *               ([.][0-9]{1,2})?     Matches if numeric after the d.p (optional)
 *               Complete Regex: [0-9]+([.][0-9]{2})?
 *
 */
@JvmInline
value class TransactionAmount(val value: Double)
{
    companion object
    {
        fun create(raw: String): Result<TransactionError, TransactionAmount>
        {
            // Pre-trim and compile regex
            val trimmed = raw.trim()
            val regex = Regex("[0-9]+(\\.[0-9]{1,2})?")

            return when
            {
                // Empty strings
                trimmed.isBlank() ->
                    Result.Failure(TransactionError.Empty)

                // Doesn't match pattern
                !trimmed.matches(regex)
                    -> Result.Failure(TransactionError.BadFormat)

                // Parse to double & catch unexpected errors
                else ->
                {
                    /*
                    Returns the TransactionAmount(Double) if non-null, or
                    TransactionError.BadFormat if it is null.
                    ?.let performs a safe call to TransactionAmount([the double])
                    "Safe" meaning only if toDoubleOrNull returned non-null

                    if the left-hand value ^ is null, ?: returns the right-hand value
                     */
                    trimmed.toDoubleOrNull()
                        ?.let { Result.Success(TransactionAmount(it)) }
                        ?: Result.Failure(TransactionError.BadFormat)
                }
            }
        }
    }
}

/**
 * Value class for category foreign keys. Prevents transactions being created
 * without a category.
 */
@JvmInline
value class TransactionCategory(val value: Long /* Foreign key */)
{
    companion object
    {
        fun create(raw: Long): Result<TransactionError, TransactionCategory>
        {
            return when
            {
                raw < 0 -> Result.Failure(TransactionError.NoCategory)
                else -> Result.Success(TransactionCategory(raw))
            }
        }
    }
}


/**
 * Value class enforcing correct date-time format
 */
@JvmInline
value class TransactionDateTime(val calendar: Calendar)
{
    companion object
    {
        fun create(rawDate: String, rawTime: String)
        {
            val dateFormat = SimpleDateFormat("dd/MM/YYYY HH:mm", Locale.getDefault())

            try
            {
                val parsedDateTime = dateFormat.parse(rawDate + " " )
            }
        }

    }
}


